File: src/open/gridLayer.ts
import { Pbf as Protobuf } from 'pbf-ts';
import { deltaDecodeArray, deltaEncodeArray } from '../util';

import type { Extents } from '.';

/** Grid object to read from */
export class GridData {
  #pbf: Protobuf;
  name: string = 'default';
  private dataPos = 0;
  extent: Extents = 8_192;
  size: number = 0;
  min: number = 0;
  max: number = 0;

  /**
   * @param pbf - the pbf protocol we are reading from
   * @param end - the position to stop at
   */
  constructor(pbf: Protobuf, end: number) {
    this.#pbf = pbf;
    pbf.readFields(this.#readGrid, this, end);
  }

  /**
   * Get the raw gridded data. Reads much like a raw image.
   * @returns - the decomposed gridded data
   */
  data(): number[] {
    this.#pbf.pos = this.dataPos;
    return deltaDecodeArray(this.#pbf.readPackedVarint()).map((v) =>
      unmapValue(v, this.min, this.max, this.extent),
    );
  }

  /**
   * @param tag - the tag to read
   * @param grid - the elevation data to mutate
   * @param pbf - the Protobuf to pull the appropriate data from
   */
  #readGrid(tag: number, grid: GridData, pbf: Protobuf): void {
    if (tag === 0) grid.extent = pbf.readVarint() as Extents;
    else if (tag === 1) grid.size = pbf.readVarint();
    else if (tag === 2) grid.min = pbf.readFloat();
    else if (tag === 3) grid.max = pbf.readFloat();
    else if (tag === 4) grid.dataPos = pbf.pos;
    else if (tag === 5) grid.name = pbf.readString();
  }
}

/**
 * Gridded data input. It is assumed data is the actual grid data of each point in the tile
 */
export interface GridInput {
  /** The name of the grid data */
  name: string;
  /** The size of the square tile. The length and width MUST be the same size */
  size: number;
  /**
   * The actual grid data assumed to be floating point (32-bit) precise.
   * It is assumed to find data you can use: `index = y * size + x`
   */
  data: number[];
  /** The extent defines the 0->extent range to remap the data to  */
  extent: Extents;
}

/**
 * @param input - the data to encode
 * @returns - the encoded data
 */
export function writeGridData(input: GridInput): Uint8Array {
  const pbf = new Protobuf();

  const { name, data, size, extent } = input;
  const max = data.reduce((a, b) => Math.max(a, b), 0);
  const min = data.reduce((a, b) => Math.min(a, b), 0);
  const reMapped = data.map((v) => remapValue(v, min, max, extent));
  const d_coded = deltaEncodeArray(reMapped);

  pbf.writeVarintField(0, extent);
  pbf.writeVarintField(1, size);
  pbf.writeFloatField(2, min);
  pbf.writeFloatField(3, max);
  pbf.writePackedVarint(4, d_coded);
  pbf.writeStringField(5, name);

  return pbf.commit();
}

/**
 * @param value - input value to remap
 * @param min - min value
 * @param max - max value
 * @param extent - extent defines the 0->extent range to remap to
 * @returns - remapped value
 */
function remapValue(value: number, min: number, max: number, extent: number): number {
  return Math.round(((value - min) * extent) / (max - min));
}

/**
 * @param value - input value to unmap
 * @param min - min value
 * @param max - max value
 * @param extent - extent defines the 0->extent range to remap from
 * @returns - unremapped value
 */
function unmapValue(value: number, min: number, max: number, extent: number): number {
  return (value * (max - min)) / extent + min;
}

/**
 * @param r - red
 * @param g - green
 * @param b - blue
 * @returns - elevation
 */
export function convertTerrariumElevationData(r: number, g: number, b: number): number {
  return r * 256.0 + g + b / 256.0 - 32768.0;
}

/**
 * @param r - red
 * @param g - green
 * @param b - blue
 * @returns - elevation
 */
export function convertMapboxElevationData(r: number, g: number, b: number): number {
  return -10000 + (r * 256 * 256 + g * 256 + b) * 0.1;
}

File: rust/open/grid_layer.rs
use crate::{delta_decode_array, delta_encode_array, open::Extent};
use alloc::{string::String, vec::Vec};
use libm::round;
use pbf::{ProtoRead, ProtoWrite, Protobuf};

// TODO: This could be faster if we don't read in the grid data on parsing but only if the user needs it

/// Gridded data object to read from
#[derive(Default, Debug, PartialEq)]
pub struct GridData {
    /// The name of the gridded data
    pub name: String,
    /// The grid data
    pub data: Vec<f64>,
    /// The extent for remapping the data with a value between 0 and extent
    pub extent: Extent,
    /// The size of the tile (width and height)
    pub size: f64,
    /// The minimum grid value
    pub min: f64,
    /// The maximum grid value
    pub max: f64,
}
impl GridData {
    /// create a new GridData object
    pub fn new(
        name: String,
        extent: Extent,
        size: f64,
        min: f64,
        max: f64,
        data: Vec<f64>,
    ) -> Self {
        GridData { name, data, extent, size, min, max }
    }
}
impl ProtoRead for GridData {
    fn read(&mut self, tag: u64, pb: &mut Protobuf) {
        match tag {
            1 => self.extent = pb.read_varint(),
            2 => self.size = pb.read_varint(),
            3 => self.min = pb.read_varint(),
            4 => self.max = pb.read_varint(),
            5 => {
                self.data = delta_decode_array(&pb.read_packed())
                    .into_iter()
                    .map(|v| unmap_value(v as f64, self.min, self.max, self.extent.into()))
                    .collect()
            }
            6 => self.name = pb.read_string(),
            _ => panic!("unknown tag: {}", tag),
        }
    }
}
impl ProtoWrite for GridData {
    fn write(&self, pb: &mut Protobuf) {
        let max = self.data.iter().fold(f64::MIN, |a, b| f64::max(a, *b));
        let min = self.data.iter().fold(f64::MAX, |a, b| f64::min(a, *b));
        let re_mapped: Vec<u32> =
            self.data.iter().map(|v| remap_value(*v, min, max, self.extent.into())).collect();
        let d_coded = delta_encode_array(&re_mapped);

        pb.write_varint_field(1, self.extent);
        pb.write_varint_field(2, self.size);
        pb.write_varint_field(3, min);
        pb.write_varint_field(4, max);
        pb.write_packed_varint(5, &d_coded);
        pb.write_string_field(6, &self.name);
    }
}

/// map the value to the range 0->extent
fn remap_value(value: f64, min: f64, max: f64, extent: f64) -> u32 {
    round(((value - min) * extent) / (max - min)) as u32
}

/// map the value back to floats
fn unmap_value(value: f64, min: f64, max: f64, extent: f64) -> f64 {
    (value * (max - min)) / extent + min
}

/// convert rgb to elevation using terrarium formula
pub fn convert_terrarium_elevation_data(r: u8, g: u8, b: u8) -> f64 {
    r as f64 * 256. + g as f64 + b as f64 / 256. - 32768.
}

/// convert rgb to elevation using mapbox formula
pub fn convert_mapbox_elevation_data(r: u8, g: u8, b: u8) -> f64 {
    -10000. + (r as f64 * 256. * 256. + g as f64 * 256. + b as f64) * 0.1
}
